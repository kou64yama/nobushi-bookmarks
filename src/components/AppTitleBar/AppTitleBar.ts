import Vue from 'vue';
import { mapActions, mapGetters, mapState } from 'vuex';

export default Vue.extend({
  computed: {
    ...mapGetters('app', ['platform']),
    ...mapState('document', ['title']),
    ...mapState('auth', ['currentUser']),
  },
  methods: {
    ...mapActions('app', ['close']),
    ...mapActions('auth', ['signOut']),
  },
});
