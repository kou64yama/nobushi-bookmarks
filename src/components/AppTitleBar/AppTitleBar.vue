<template>
  <v-system-bar
    app
    dark
    window
    color="primary"
    class="title-bar"
  >
    <v-spacer/>
    {{ title }}
    <v-spacer/>
    <v-menu offset-y>
      <v-btn :disabled="!currentUser" color="red" class="btn mr-4" style="min-width: initial;" slot="activator">
        <v-icon class="mr-0">person</v-icon>
      </v-btn>
      <v-list two-line class="py-0" v-if="currentUser">
        <v-list-tile>
          <v-list-tile-avatar color="grey lighten-1">
            <img :src="currentUser.photoURL" v-if="currentUser.photoURL">
            <v-icon color="white" v-else>person</v-icon>
          </v-list-tile-avatar>
          <v-list-tile-content>
            <v-list-tile-title>
              {{ currentUser.displayName || currentUser.email || currentUser.uid }}
            </v-list-tile-title>
            <v-list-tile-sub-title>
              {{ currentUser.email }}
            </v-list-tile-sub-title>
          </v-list-tile-content>
        </v-list-tile>
      </v-list>
      <v-divider/>
      <v-list class="py-0">
        <v-list-tile ref="signOut" @click="signOut">
          <v-list-tile-action>
            <v-icon>exit_to_app</v-icon>
          </v-list-tile-action>
          <v-list-tile-content>
            <v-list-tile-title>Sign out</v-list-tile-title>
          </v-list-tile-content>
        </v-list-tile>
      </v-list>
    </v-menu>
    <v-icon ref="close" class="btn" @click="close" v-if="platform !== 'darwin'">
      close
    </v-icon>
  </v-system-bar>
</template>

<script lang="ts" src="./AppTitleBar.ts"></script>

<style scoped>
.title-bar {
  -webkit-app-region: drag;
  user-select: none;
}

.btn {
  -webkit-app-region: no-drag;
  cursor: pointer;
}
</style>
