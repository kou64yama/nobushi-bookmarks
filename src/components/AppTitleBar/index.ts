import AppTitleBar from './AppTitleBar.vue';

export default AppTitleBar;
