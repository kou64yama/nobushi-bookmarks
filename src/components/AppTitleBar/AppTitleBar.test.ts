import Vue from 'vue';
import { mount, Stubs } from '@vue/test-utils';
import AppTitleBar from './AppTitleBar.vue';

const stubs: Stubs = {
  VSystemBar: true,
  VSpacer: true,
  VMenu: true,
  VBtn: true,
  VList: true,
  VListTile: true,
  VListTileAvatar: true,
  VListTileContent: true,
  VListTileTitle: true,
  VListTileSubTitle: true,
  VListTileAction: true,
  VIcon: true,
  VDivider: true,
};

const currentUser = {
  displayName: 'mockDisplayName',
  email: 'mockEmail',
  phoneNumber: 'mockPhoneNumber',
  photoURL: 'mockPhotoURL',
  providerId: 'mockProviderId',
  uid: 'mockUid',
};

describe('AppTitleBar', () => {
  it('should render correctly with currentUser', () => {
    const wrapper = mount(AppTitleBar, {
      stubs,
      computed: {
        platform: () => 'win32',
        title: () => 'Test Title',
        currentUser: () => currentUser,
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  it('should render correctly without currentUser', () => {
    const wrapper = mount(AppTitleBar, {
      stubs,
      computed: {
        platform: () => 'win32',
        title: () => 'Test Title',
        currentUser: () => null,
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  it('should render close button when platform is win32', () => {
    const wrapper = mount(AppTitleBar, {
      stubs,
      computed: {
        platform: () => 'win32',
        title: () => 'Test Title',
        currentUser: () => currentUser,
      },
    });

    expect(wrapper.vm.$refs.close).toBeDefined();
  });

  it('should not render close button when platform is darwin', () => {
    const wrapper = mount(AppTitleBar, {
      stubs,
      computed: {
        platform: () => 'darwin',
        title: () => 'Test Title',
        currentUser: () => currentUser,
      },
    });

    expect(wrapper.vm.$refs.close).toBeUndefined();
  });

  it('should call close when close button is clicked', () => {
    const close = jest.fn();
    const wrapper = mount(AppTitleBar, {
      stubs,
      computed: {
        platform: () => 'win32',
        title: () => 'Test Title',
        currentUser: () => currentUser,
      },
      methods: {
        close,
      },
    });

    (wrapper.vm.$refs.close as Vue).$emit('click');

    expect(close).toBeCalled();
  });

  it('should call signOut when sign out list item is clicked', () => {
    const signOut = jest.fn();
    const wrapper = mount(AppTitleBar, {
      stubs,
      computed: {
        platform: () => 'win32',
        title: () => 'Test Title',
        currentUser: () => currentUser,
      },
      methods: {
        signOut,
      },
    });

    (wrapper.vm.$refs.signOut as Vue).$emit('click');

    expect(signOut).toBeCalled();
  });
});
