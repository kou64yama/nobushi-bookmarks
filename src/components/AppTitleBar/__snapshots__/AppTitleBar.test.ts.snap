// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppTitleBar should render correctly with currentUser 1`] = `
<vsystembar-stub
  app=""
  class="title-bar"
  color="primary"
  dark=""
  window=""
>
  <vspacer-stub />
  
  Test Title
  
  <vspacer-stub />
   
  <vmenu-stub
    offset-y=""
  >
    <vbtn-stub
      class="btn mr-4"
      color="red"
      style="min-width: initial;"
    >
      <vicon-stub
        class="mr-0"
      >
        person
      </vicon-stub>
    </vbtn-stub>
     
    <vlist-stub
      class="py-0"
      two-line=""
    >
      <vlisttile-stub>
        <vlisttileavatar-stub
          color="grey lighten-1"
        >
          <img
            src="mockPhotoURL"
          />
        </vlisttileavatar-stub>
         
        <vlisttilecontent-stub>
          <vlisttiletitle-stub>
            
            mockDisplayName
          
          </vlisttiletitle-stub>
           
          <vlisttilesubtitle-stub>
            
            mockEmail
          
          </vlisttilesubtitle-stub>
        </vlisttilecontent-stub>
      </vlisttile-stub>
    </vlist-stub>
     
    <vdivider-stub />
     
    <vlist-stub
      class="py-0"
    >
      <vlisttile-stub>
        <vlisttileaction-stub>
          <vicon-stub>
            exit_to_app
          </vicon-stub>
        </vlisttileaction-stub>
         
        <vlisttilecontent-stub>
          <vlisttiletitle-stub>
            Sign out
          </vlisttiletitle-stub>
        </vlisttilecontent-stub>
      </vlisttile-stub>
    </vlist-stub>
  </vmenu-stub>
   
  <vicon-stub
    class="btn"
  >
    
    close
  
  </vicon-stub>
</vsystembar-stub>
`;

exports[`AppTitleBar should render correctly without currentUser 1`] = `
<vsystembar-stub
  app=""
  class="title-bar"
  color="primary"
  dark=""
  window=""
>
  <vspacer-stub />
  
  Test Title
  
  <vspacer-stub />
   
  <vmenu-stub
    offset-y=""
  >
    <vbtn-stub
      class="btn mr-4"
      color="red"
      disabled="true"
      style="min-width: initial;"
    >
      <vicon-stub
        class="mr-0"
      >
        person
      </vicon-stub>
    </vbtn-stub>
     
    <!---->
     
    <vdivider-stub />
     
    <vlist-stub
      class="py-0"
    >
      <vlisttile-stub>
        <vlisttileaction-stub>
          <vicon-stub>
            exit_to_app
          </vicon-stub>
        </vlisttileaction-stub>
         
        <vlisttilecontent-stub>
          <vlisttiletitle-stub>
            Sign out
          </vlisttiletitle-stub>
        </vlisttilecontent-stub>
      </vlisttile-stub>
    </vlist-stub>
  </vmenu-stub>
   
  <vicon-stub
    class="btn"
  >
    
    close
  
  </vicon-stub>
</vsystembar-stub>
`;
